pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test: TestTenantLogic_guardrails failed

pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test.TestTenantLogic_guardrails [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_ScratchProjectPutTcExperimentsInHere_SqlLogicDeclarativeSchemaChangerCorpusBazel/10578523?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_ScratchProjectPutTcExperimentsInHere_SqlLogicDeclarativeSchemaChangerCorpusBazel/10578523?buildTab=artifacts#/) on release-23.1 @ [5b0d9c6abb00a2ffac00ad75c3bea30577126f05](https://github.com/cockroachdb/cockroach/commits/5b0d9c6abb00a2ffac00ad75c3bea30577126f05):


```
=== RUN   TestTenantLogic_guardrails
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/5361bfdf8ed35964b6d64e8226ea4e53/logTestTenantLogic_guardrails195015607
    test_log_scope.go:79: use -show-logs to present logs inline
[05:52:30] setting distsql_workmem='51781B';
[05:52:30] setting distsql_workmem='51781B';
    logic.go:2886: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3937/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test_/3node-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/guardrails:34: SELECT * FROM guardrails
        expected:
        txn has read 51 rows, which is above the limit
        got:
        pq: txn has read 52 rows, which is above the limit: TxnID 8123e6db-4efe-435f-bfc8-ef11a2e29fe8 SessionID 1769aaffa47a435a0000000000000001
[05:52:30] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3937/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test_/3node-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/guardrails with config 3node-tenant: 9 tests, 1 failures
    logic.go:3998: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3937/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test_/3node-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/guardrails:37: error while processing
    logic.go:3998: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3937/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test_/3node-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/guardrails:37: too many errors encountered, skipping the rest of the input
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5361bfdf8ed35964b6d64e8226ea4e53/logTestTenantLogic_guardrails195015607
--- FAIL: TestTenantLogic_guardrails (12.44s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantLogic_guardrails.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-28874

From the log:
```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //pkg/ccl/logictestccl/tests/3node-tenant:3node-tenant_test
-----------------------------------------------------------------------------
initialized metamorphic constant "span-reuse-rate" with value 7
initialized metamorphic constant "parse-json-impl" with value 1
initialized metamorphic constant "storage.ingest_as_flushable.enabled" with value false
initialized metamorphic constant "mvcc-max-iters-before-seek" with value 2
initialized metamorphic constant "default-batch-bytes-limit" with value 22641
initialized metamorphic constant "kv-batch-size" with value 1
initialized metamorphic constant "row-container-rows-per-chunk-shift" with value 1
initialized metamorphic constant "inverted-joiner-batch-size" with value 1
initialized metamorphic constant "spilling-queue-initial-len" with value 15
initialized metamorphic constant "direct-scans-enabled" with value true
initialized metamorphic constant "ColIndexJoin-batch-size" with value 480824
initialized metamorphic constant "ColIndexJoin-using-streamer-batch-size" with value 8288613
initialized metamorphic constant "max-batch-size" with value 6919
initialized metamorphic constant "max-batch-byte-size" with value 21027001
initialized metamorphic constant "parallel-scan-result-threshold" with value 2760
initialized metamorphic constant "copy-batch-size" with value 13024
initialized metamorphic constant "async-IE-result-channel-buffer-size" with value 23
initialized metamorphic constant "use-index-lookup-for-descriptors-in-database" with value false
initialized metamorphic constant "copy-fast-path-enabled-default" with value false
initialized metamorphic constant "backup-merge-file-buffer-size" with value 5363147
initialized metamorphic constant "write-metadata-sst" with value true
initialized metamorphic constant "backup.write_metadata_with_external_ssts.enabled" with value false
initialized metamorphic constant "raft-log-truncation-clearrange-threshold" with value 418428
initialized metamorphic constant "restore-worker-concurrency" with value 1
initialized metamorphic constant "changefeed.event_consumer_worker_queue_size" with value 8
initialized metamorphic constant "changefeed.balance_range_distribution.enable" with value true
initialized metamorphic constant "changefeed.fast_gzip.enabled" with value false
initialized metamorphic constant "changefeed.new_webhook_sink_enabled" with value true
initialized metamorphic constant "logictest-global-mvcc-range-tombstone" with value true
initialized metamorphic constant "logictest-use-mvcc-range-tombstones-for-point-deletes" with value true
initialized metamorphic constant "parse-json-impl" with value 0
initialized metamorphic constant "use-mux-rangefeed" with value true
initialized metamorphic constant "mvcc-incremental-iter-tbi" with value false
initialized metamorphic constant "mvcc-value-disable-simple-encoding" with value true
initialized metamorphic constant "storage.value_blocks.enabled" with value false
initialized metamorphic constant "storage.ingest_as_flushable.enabled" with value false
initialized metamorphic constant "default-batch-bytes-limit" with value 945
initialized metamorphic constant "datum-row-converter-batch-size" with value 1
initialized metamorphic constant "row-container-rows-per-chunk-shift" with value 1
initialized metamorphic constant "inverted-joiner-batch-size" with value 1
initialized metamorphic constant "zig-zag-joiner-batch-size" with value 1
initialized metamorphic constant "merge-joiner-groups-buffer" with value 8
initialized metamorphic constant "direct-scans-enabled" with value true
initialized metamorphic constant "ColIndexJoin-using-streamer-batch-size" with value 5244296
initialized metamorphic constant "max-batch-size" with value 5192
initialized metamorphic constant "max-batch-byte-size" with value 27297691
initialized metamorphic constant "resolve-membership-single-scan-enabled" with value false
initialized metamorphic constant "copy-batch-size" with value 11168
initialized metamorphic constant "async-IE-result-channel-buffer-size" with value 2
initialized metamorphic constant "copy-fast-path-enabled-default" with value false
initialized metamorphic constant "write-metadata-sst" with value true
initialized metamorphic constant "raft-log-truncation-clearrange-threshold" with value 212166
initialized metamorphic constant "backup-read-manifest" with value true
initialized metamorphic constant "restore-worker-concurrency" with value 4
initialized metamorphic constant "changefeed.mux_rangefeed.enabled" with value true
initialized metamorphic constant "changefeed.event_consumer_worker_queue_size" with value 0
initialized metamorphic constant "changefeed.balance_range_distribution.enable" with value true
initialized metamorphic constant "changefeed.fast_gzip.enabled" with value false
initialized metamorphic constant "changefeed.new_pubsub_sink_enabled" with value true
initialized metamorphic constant "aws-reuse-kms" with value false
initialized metamorphic constant "logictest-use-mvcc-range-tombstones-for-point-deletes" with value true
I230618 05:50:28.548139 1 (gostd) rand.go:249  [-] 1  random seed: 2741483707705172504
...
=== RUN   TestTenantLogic_guardrails
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/5361bfdf8ed35964b6d64e8226ea4e53/logTestTenantLogic_guardrails195015607
    test_log_scope.go:79: use -show-logs to present logs inline
[05:52:30] setting distsql_workmem='51781B';
[05:52:30] setting distsql_workmem='51781B';
    logic.go:2886: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3937/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test_/3node-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/guardrails:34: SELECT * FROM guardrails
        expected:
        txn has read 51 rows, which is above the limit
        got:
        pq: txn has read 52 rows, which is above the limit: TxnID 8123e6db-4efe-435f-bfc8-ef11a2e29fe8 SessionID 1769aaffa47a435a0000000000000001
[05:52:30] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3937/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test_/3node-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/guardrails with config 3node-tenant: 9 tests, 1 failures
    logic.go:3998: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3937/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test_/3node-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/guardrails:37: error while processing
    logic.go:3998: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3937/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test_/3node-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/guardrails:37: too many errors encountered, skipping the rest of the input
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5361bfdf8ed35964b6d64e8226ea4e53/logTestTenantLogic_guardrails195015607
--- FAIL: TestTenantLogic_guardrails (12.44s)
```

--- a/pkg/sql/logictest/BUILD.bazel
+++ b/pkg/sql/logictest/BUILD.bazel
@@ -48,7 +48,6 @@ go_library(
         "//pkg/sql/rowinfra",
         "//pkg/sql/schemachanger/corpus",
         "//pkg/sql/schemachanger/scexec",
-        "//pkg/sql/schemachanger/scplan",
         "//pkg/sql/sem/eval",
         "//pkg/sql/sem/tree",
         "//pkg/sql/sessiondatapb",

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -62,7 +62,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/rowinfra"
 	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/corpus"
 	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scexec"
-	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scplan"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb"
@@ -1338,9 +1337,30 @@ func (t *logicTest) newCluster(
 		}
 		return st
 	}
-	var corpusCollectionCallback func(p scplan.Plan, stageIdx int) error
-	if serverArgs.DeclarativeCorpusCollection && t.declarativeCorpusCollector != nil {
-		corpusCollectionCallback = t.declarativeCorpusCollector.GetBeforeStage(t.rootT.Name(), t.t())
+	setSQLTestingKnobs := func(knobs *base.TestingKnobs) {
+		knobs.SQLEvalContext = &eval.TestingKnobs{
+			AssertBinaryExprReturnTypes:     true,
+			AssertUnaryExprReturnTypes:      true,
+			AssertFuncExprReturnTypes:       true,
+			DisableOptimizerRuleProbability: *disableOptRuleProbability,
+			OptimizerCostPerturbation:       *optimizerCostPerturbation,
+			ForceProductionValues:           serverArgs.ForceProductionValues,
+		}
+		knobs.SQLExecutor = &sql.ExecutorTestingKnobs{
+			DeterministicExplain:            true,
+			UseTransactionalDescIDGenerator: true,
+		}
+		knobs.SQLStatsKnobs = &sqlstats.TestingKnobs{
+			AOSTClause: "AS OF SYSTEM TIME '-1us'",
+		}
+		if serverArgs.DeclarativeCorpusCollection && t.declarativeCorpusCollector != nil {
+			knobs.SQLDeclarativeSchemaChanger = &scexec.TestingKnobs{
+				BeforeStage: t.declarativeCorpusCollector.GetBeforeStage(t.rootT.Name(), t.t()),
+			}
+		}
+		knobs.DistSQL = &execinfra.TestingKnobs{
+			ForceDiskSpill: t.cfg.SQLExecUseDisk,
+		}
 	}
 	// TODO(andrei): if createTestServerParams() is used here, the command filter
 	// it installs detects a transaction that doesn't have
@@ -1377,24 +1397,6 @@ func (t *logicTest) newCluster(
 						UseRangeTombstonesForPointDeletes: shouldUseMVCCRangeTombstonesForPointDeletes,
 					},
 				},
-				SQLEvalContext: &eval.TestingKnobs{
-					AssertBinaryExprReturnTypes:     true,
-					AssertUnaryExprReturnTypes:      true,
-					AssertFuncExprReturnTypes:       true,
-					DisableOptimizerRuleProbability: *disableOptRuleProbability,
-					OptimizerCostPerturbation:       *optimizerCostPerturbation,
-					ForceProductionValues:           serverArgs.ForceProductionValues,
-				},
-				SQLExecutor: &sql.ExecutorTestingKnobs{
-					DeterministicExplain:            true,
-					UseTransactionalDescIDGenerator: true,
-				},
-				SQLStatsKnobs: &sqlstats.TestingKnobs{
-					AOSTClause: "AS OF SYSTEM TIME '-1us'",
-				},
-				SQLDeclarativeSchemaChanger: &scexec.TestingKnobs{
-					BeforeStage: corpusCollectionCallback,
-				},
 				RangeFeed: &rangefeed.TestingKnobs{
 					IgnoreOnDeleteRangeError: ignoreMVCCRangeTombstoneErrors,
 				},
@@ -1406,15 +1408,13 @@ func (t *logicTest) newCluster(
 		// matter where the data really is.
 		ReplicationMode: base.ReplicationManual,
 	}
+	setSQLTestingKnobs(&params.ServerArgs.Knobs)
 
 	cfg := t.cfg
 	if cfg.DefaultTestTenant == base.TestTenantEnabled {
 		// In the tenant case we need to enable replication in order to split and
 		// relocate ranges correctly.
 		params.ReplicationMode = base.ReplicationAuto
-	}
-	params.ServerArgs.Knobs.DistSQL = &execinfra.TestingKnobs{
-		ForceDiskSpill: cfg.SQLExecUseDisk,
 	}
 	if cfg.BootstrapVersion != clusterversion.Key(0) {
 		if params.ServerArgs.Knobs.Server == nil {
@@ -1490,13 +1490,6 @@ func (t *logicTest) newCluster(
 				TenantID: serverutils.TestTenantID(),
 				Settings: settings,
 				TestingKnobs: base.TestingKnobs{
-					SQLExecutor: &sql.ExecutorTestingKnobs{
-						DeterministicExplain:            true,
-						UseTransactionalDescIDGenerator: true,
-					},
-					SQLStatsKnobs: &sqlstats.TestingKnobs{
-						AOSTClause: "AS OF SYSTEM TIME '-1us'",
-					},
 					RangeFeed: paramsPerNode[i].Knobs.RangeFeed,
 				},
 				MemoryPoolSize:    params.ServerArgs.SQLMemoryPoolSize,
@@ -1506,6 +1499,7 @@ func (t *logicTest) newCluster(
 				// Give every tenant its own ExternalIO directory.
 				ExternalIODir: path.Join(t.sharedIODir, strconv.Itoa(i)),
 			}
+			setSQLTestingKnobs(&tenantArgs.TestingKnobs)
 
 			for _, opt := range knobOpts {
 				t.rootT.Logf("apply knob opt %T to tenant", opt)

